services/platform/platform.go

```diff
@@ -112,34 +117,119 @@ func (s *Service) SignUp(ctx context.Context, email, firstName, lastName string)
+		pmmServerURL:              pmmServerURL,
+		pmmServerOAuthCallbackURL: fmt.Sprintf("%s/login/generic_oauth", pmmServerURL),
+		pmmServerID:               settings.PMMServerID,
+	})
 	if err != nil {
 		return err
 	}
 
-	err = s.db.InTransaction(func(tx *reform.TX) error {
-		params := models.ChangeSettingsParams{SessionID: resp.SessionId, Email: email}
-		_, err := models.UpdateSettings(tx.Querier, &params)
-		return err
+	err = models.InsertPerconaSSODetails(s.db.Querier, &models.PerconaSSODetails{
+		ClientID:     ssoParams.ClientID,
+		ClientSecret: ssoParams.ClientSecret,
+		IssuerURL:    ssoParams.IssuerURL,
+		Scope:        ssoParams.Scope,
+	})
+	return errors.Wrap(err, "failed to save SSO details")
+}
+
+type connectPMMParams struct {
+	pmmServerURL, pmmServerOAuthCallbackURL, pmmServerID, serverName, email, password string
+}
+
+type connectPMMRequest struct {
+	PMMServerID               string `json:"pmm_server_id"`
+	PMMServerName             string `json:"pmm_server_name"`
+	PMMServerURL              string `json:"pmm_server_url"`
+	PMMServerOAuthCallbackURL string `json:"pmm_server_oauth_callback_url"`
+}
+
+type ssoDetails struct {
+	ClientID     string `json:"client_id"`
+	ClientSecret string `json:"client_secret"`
+	Scope        string `json:"scope"`
+	IssuerURL    string `json:"issuer_url"`
+}
+
+type connectPMMResponse struct {
+	SSODetails *ssoDetails `json:"sso_details"`
+}
+
+type grpcGatewayError struct {
+	Message string `json:"message"`
+	Code    uint32 `json:"code"`
+}
+
+func (s *Service) connect(ctx context.Context, params *connectPMMParams) (*ssoDetails, error) {
+	endpoint := fmt.Sprintf("https://%s/v1/orgs/inventory", s.host)
+	if params.pmmServerID == "" {
+		return nil, errors.Errorf("PMMServerID is not yet set, please wait until a service %q is added into the inventory", models.DefaultPMMPostgreSQLService)
+	}
+	marshaled, err := json.Marshal(connectPMMRequest{
+		PMMServerID:               params.pmmServerID,
+		PMMServerName:             params.serverName,
+		PMMServerURL:              params.pmmServerURL,
+		PMMServerOAuthCallbackURL: params.pmmServerOAuthCallbackURL,
 	})
 	if err != nil {
-		return errors.Wrap(err, "failed to save session id")
+		return nil, errors.Wrap(err, "failed to marshal request data")
 	}
 
-	return nil
+	client := http.Client{Timeout: time.Second * 10}
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(marshaled))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to build request")
+	}
+	encodedEmailPassword := base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%s:%s", params.email, params.password)))
+	req.Header.Add("Authorization", fmt.Sprintf("Basic %s", encodedEmailPassword))
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to execute request")
+	}
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			s.l.Errorf("Failed to close response body: %s", err)
```

Comment:
is this error very important to capture? can't we simply write a single defer statement closing the response body instead of this block?

Reply:
done